### ISSUE TITLE
Scope of namespace declarations and XPath expressions defective

### ISSUE THREAD

**Issue by jperterm** [DeltaXML/vscode-xslt-tokenizer#116]
I discovered two issues with namespace declaration scopes and XPath expressions.

First thing is a missing error, that a namespace is undeclared:

```xsl
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
                xmlns:xs="http://www.w3.org/2001/XMLSchema"
                exclude-result-prefixes="#all"
                version="3.0">
    <xsl:output method="xml" indent="yes"/>
    <xsl:mode on-no-match="shallow-copy"/>
    <xsl:template match="/*" mode="#all">
        <out>
            <xsl:value-of xmlns:ns1="urn:namespace1" select="@ns1:attr"/>
            <xsl:value-of select="@ns1:attr"/><!-- Missing error caused by line above -->
        </out>
    </xsl:template>
</xsl:stylesheet>
```

The second thing is an error that must not occur:

```xsl
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
                xmlns:xs="http://www.w3.org/2001/XMLSchema"
                exclude-result-prefixes="#all"
                version="3.0">
    <xsl:output method="xml" indent="yes"/>
    <xsl:mode on-no-match="shallow-copy"/>
    <xsl:template match="/*" mode="#all">
        <out>
            <!-- False error caused by namespace declaration **following up** the select attribute -->
            <xsl:value-of select="@ns1:attr" xmlns:ns1="urn:namespace1"/>
        </out>
    </xsl:template>
</xsl:stylesheet>
```

**Comment by pgfearo** [comment]
Even though related I will treat these 2 bugs as separate:

bug #1 (missing error) is caused by the primitive way in which in-scope namespaces are tracked - the solution is fairly trivial - I will tackle this next

bug #2 (false error because `xmlns` occurs after `select` attribute) is fixed and awaiting next release. To fix this, when an xmlns declaration is encountered, any 'prefix problems' for the declared prefix that are on the same element as the xmlns declaration are removed.

**Comment by pgfearo** [comment]
For bug(2) there was a related problem when xmlns declarations occurred after a name attribute with a prefixed value, e.g.

``` <xsl:function name="fn:test" xmlns:fn="com.deltaxml"> ```

This issue is also fixed by removing the problem when the matching xmlns on the same element is found.

**Comment by pgfearo** [comment]
Fix for bug(1) (missing error) now also fixed on the github repo - bug(1) and bug(2) fixes now awaiting release.

 this occurred when any preceding sibling element had an xmlns declaration with that prefix. The preceding sibling erroneously passing on namespace scope could be an element with start/end tags or a self-closed element.

**Comment by pgfearo** [comment]
Namespace scoping bugs fixed in Release 1.4.3
